source/common/common/utility.cc

```diff
@@ -122,6 +123,15 @@ std::vector<std::string> StringUtil::split(const std::string& source, const std:
   return ret;
 }
 
+std::string StringUtil::join(const std::vector<std::string>& source, const std::string& delimiter) {
+  std::ostringstream buf;
+  std::copy(source.begin(), source.end(),
+            std::ostream_iterator<std::string>(buf, delimiter.c_str()));
+  std::string ret = buf.str();
+  // copy will always end with an extra delimiter, we remove it here.
+  return ret.substr(0, ret.length() - delimiter.length());
```

Comment:
If the elements to join is empty, the std::copy bit still ends up adding a delimiter. It's odd, but that's how it is. I added a test for this case and the result is an empty string, as expected.